### PR TITLE
REGR: preserve freq in DTI/TDI outer join

### LINKED ISSUE #32157: API: joining DatetimeIndexes not preserving freq
```
dti = pd.date_range('2016-01-01', periods=10)
left, right = dti[:5], dti[5:]
joined = left.join(right, how="outer")
```

In master `joined.freq` comes back as None, while in 0.25.3 it was "D".  There is some code in DatetimeIndex._wrap_joined_index that suggests to me this was intentional.

Can someone confirm this suspicion cc @jorisvandenbossche ?  If so, I've got a branch that restores the old behavior, just need to add tests.

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #32157
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

also just a nice cleanup, hoping to get this into ExtensionIndex before long

### CODE REVIEW

File: pandas/tests/indexes/datetimes/test_setops.py [hunk 1]
```diff
@@ -367,6 +367,19 @@ def test_join_nonunique(self):
         rs = idx1.join(idx2, how="outer")
         assert rs.is_monotonic
 
+    @pytest.mark.parametrize("tz", [None, "US/Pacific"])
```
Comment (mroeschke) [inline pandas/tests/indexes/datetimes/test_setops.py]: Could we use the `tz_aware_fixture` here?

Comment (jbrockmendel) [reply pandas/tests/indexes/datetimes/test_setops.py]: sure

**Comment by jreback** [comment]
i marked for 1.1, but is this a regression?

**Comment by jreback** [comment]
rebase as well

**Comment by jbrockmendel** [comment]
> i marked for 1.1, but is this a regression?

The behavior changed between 0.25.3 and 1.0, but it isnt obvious whether to call it a regression or not because it didnt obviously break anything. (see #32157)

**Comment by jreback** [comment]
> > i marked for 1.1, but is this a regression?
> 
> The behavior changed between 0.25.3 and 1.0, but it isnt obvious whether to call it a regression or not because it didnt obviously break anything. (see #32157)

ok it was marked originally as a regression
however i bet this will non trivial to backport (tests)

so up to u

**Comment by jbrockmendel** [comment]
removed REGR tag

**Comment by jreback** [comment]
hmm i think this needs a whatsnew note (to document that this is changing), otherwise lgtm.

File: doc/source/whatsnew/v1.1.0.rst [hunk 2]
```diff
@@ -57,6 +57,7 @@ Other API changes
   will now be ``min`` and ``max`` to match with numeric dtypes in :meth:`DataFrame.describe` (:issue:`30164`)
 - :meth:`Groupby.groups` now returns an abbreviated representation when called on large dataframes (:issue:`1135`)
 - ``loc`` lookups with an object-dtype :class:`Index` and an integer key will now raise ``KeyError`` instead of ``TypeError`` when key is missing (:issue:`31905`)
+- Joining on :class:`DatetimeIndex` or :class:`TimedeltaIndex` will preserve ``freq`` in simple cases (:issue:`32166`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.1.0.rst]: i guess need to backport, can you move to 1.02

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @jbrockmendel

**Comment by jreback** [comment]
@jbrockmendel as expected this needs a manual backport :-<

**Pull request merged**